### DISCUSSION TITLE
`ftp` がリンクのような扱いになる

### DISCUSSION THREAD

**Discussion by yarakos95** [increments/qiita-discussions#893] [要望・アイデア・不具合]
Qiita では、`https` に続くリンクはオートリンクかリンクカードとして表示されるようになっています。

この内、URI が `ftp` のものも `https` のリンクかのように色が変わり下線が引かれます。

これは奇妙であり、ただの文字列として表示されるべきものだと思います。

![image](https://github.com/user-attachments/assets/bf515d77-bd8f-4e4b-987c-52930e3258c5)

```markdown
## 独立

ftp://example.com

http://example.com

https://example.com

## 文章内

文章内の ftp://example.com

文章内の http://example.com

文章内の https://example.com

```

**Comment by tanaka-asami415** [comment]
@Yarakashi-Kikohshi 
ご意見ありがとうございます！

Qiitaで利用している[Markdownパーサー](https://help.qiita.com/ja/articles/qiita-markdown) の仕様である可能性が高そうですが、一度調査してみます🙇

**Reply by ohakutsu** [reply]
@Yarakashi-Kikohshi 
おまたせしました！

`ftp`がリンクになるのは、Qiitaで利用しているMarkdown（GitHub Flavored Markdown）の仕様の様でした。
現時点では、GitHub Flavored Markdownの仕様にできる限り準拠する方針で考えております。

ですが、ご要望が多ければ改めて検討したいと考えておりますので、こちらのDiscussions等で他のユーザーさまからのご意見もお待ちして検討したいと思います🙇

**Reply by yarakos95** [reply]
@ohakutsu 様

> 現時点では、GitHub Flavored Markdownの仕様にできる限り準拠する方針で考えております。
>
> ですが、ご要望が多ければ改めて検討したいと考えておりますので、

GFM における Autolinks の仕様で `ftp` もリンクであるとの理解でしょうか。

[Autolinks][] の仕様には以下のように記述されています。

[autolinks]: https://github.github.com/gfm/#autolinks-extension- "§ 6.8Autolinks - GitHub Flavored Markdown Spec"

> An [extended www autolink](https://github.github.com/gfm/#extended-www-autolink) will be recognized when the text `www.` is found followed by a [valid domain](https://github.github.com/gfm/#valid-domain). A [valid domain](https://github.github.com/gfm/#valid-domain) consists of segments of alphanumeric characters, underscores (`_`) and hyphens (`-`) separated by periods (`.`). There must be at least one period, and no underscores may be present in the last two segments of the domain.

> An [extended url autolink](https://github.github.com/gfm/#extended-url-autolink) will be recognised when one of the schemes `http://`, or `https://`, followed by a [valid domain](https://github.github.com/gfm/#valid-domain), then zero or more non-space non-`<` characters according to [extended autolink path validation](https://github.github.com/gfm/#extended-autolink-path-validation):

これらを見る限り、「`www.`、`http://`、`https://` から始まるいずれかの場合のときのみにリンクとして解釈される」 となっていると思います。この点を踏まえても、`ftp` がリンクのように扱われるのはやはり疑問が残ります。

また、実際にここで `ftp` から始まるものを書いてもリンクとして解釈されていません。

ftp://example.com

文章内の ftp://example.com

**Reply by P-SiZK** [reply]
@Yarakashi-Kikohshi
ご返信頂きありがとうございます！

> これらを見る限り、「`www.`、`http://`、`https://` から始まるいずれかの場合のときのみにリンクとして解釈される」 となっていると思います。

GFM のドキュメントと思われるものが下記の通り、複数存在しているようです。
- https://github.github.com/gfm/
- https://github.com/github/cmark-gfm/blob/0.29.0.gfm.13/test/spec.txt

GFM の仕様がどのドキュメントに依拠しているのか次第にはなるのですが、
`ftp` のリンクの挙動についても検討したいと思います。

また GitHub 上でリンクとして解釈されないことに関しては、推測になるのですが GFM をカスタムしたものを使っているためだと思われます。

**Reply by yarakos95** [reply]
@P-SiZK 様

> GFM のドキュメントと思われるものが下記の通り、複数存在しているようです。

これは衝撃的でした。
GitHub\.com 内で利用できる Markdown 記法の仕様が GFM をカスタムしたものである [^1] ことには同意するのですが、GFM 仕様そのものにバリエーションがあることは知りませんでした。（2 つの GFM には他に違いがあるのかしら）

[^1]: GFM に記載されていない Markdown 記法やサニタイズされる HTML が多数ある

Qiita Markdown が私の前述した GFM spec ではなく [cmark-gfm][] の仕様に従うとするのであれば、確かに `ftp` がリンクとして解釈されることも矛盾しませんね。

[cmark-gfm]: https://github.com/github/cmark-gfm/tree/0.29.0.gfm.13 "github/cmark-gfm at 0.29.0.gfm.13"

知らなかったとは言え早計でした。ご回答いただきありがとうございます。

> `ftp` のリンクの挙動についても検討したいと思います。

一方で、特定の仕様に従っているにしても `ftp` がリンクのように扱われているのは奇妙だと感じているので、検討を継続していただけるとありがたいです。

**Reply by radian-jp** [reply]
昔はブラウザでFTP URLも開けたので、その辺りの名残ではないかと思われます。
現在のChromeでは廃止されているので、追従していくのがいいかもしれませんね。
([Chrome 86: Improved Focus Highlighting, WebHID, and More](https://blog.chromium.org/2020/09/chrome-86-improved-focus-highlighting.html) の Deprecate FTP Support 辺り)
